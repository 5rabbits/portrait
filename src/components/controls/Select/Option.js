/* eslint-disable */

import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import tinycolor from 'tinycolor2'
import cx from 'classnames'

export default class Option extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    color: PropTypes.string,
    focused: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    onMouseEnter: PropTypes.func.isRequired,
    option: PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    }).isRequired,
    optionRef: PropTypes.func.isRequired,
    selected: PropTypes.bool,
    style: PropTypes.object,
  }

  static defaultProps = {
    color: undefined,
    focused: false,
    selected: false,
    style: {},
  }

  handleClick = () => {
    this.props.onClick(this.props.option)
  }

  handleMouseEnter = () => {
    this.props.onMouseEnter(this.props.option)
  }

  optionRef = node => {
    this.props.optionRef(this.props.option, node)
  }

  render() {
    const { children, color, focused, selected } = this.props
    const style = { ...this.props.style }

    if (selected && color) {
      style.backgroundColor = color
    }
    else if (focused && color) {
      style.backgroundColor = tinycolor(color).lighten(35).toHexString()
    }

    if (style.backgroundColor) {
      style.color = tinycolor.mostReadable(style.backgroundColor, ['#697D91', '#FFF'])
    }

    return (
      <div
        className={cx('Select__option', {
          'Select__option--focused': focused,
        })}
        onClick={this.handleClick}
        onMouseEnter={this.handleMouseEnter}
        ref={this.optionRef}
        style={style}
        >
        <div className="Select__option__text">
          {children}
        </div>
        {selected &&
          <div className="Select__option__icon">
            <svg viewBox="0 0 400 400">
              <path d="M397.692 58.953l-52.8-41.645c-1.636-1.346-3.656-2.308-5.964-2.308-2.31 0-4.425.962-6.06 2.404l-192.065 246.79s-75.5-72.613-77.615-74.73c-2.116-2.115-4.905-5.673-9.137-5.673-4.23 0-6.154 2.982-8.366 5.194-1.635 1.73-28.565 30.007-41.837 44.05-.77.865-1.25 1.346-1.923 2.02C.77 236.687 0 238.513 0 240.533c0 2.117.77 3.848 1.924 5.483l2.692 2.5S138.59 377.203 140.803 379.416c2.212 2.212 4.905 5 8.848 5 3.848 0 7.022-4.135 8.85-5.962L398.076 70.686C399.23 69.05 400 67.224 400 65.108c0-2.404-.962-4.424-2.308-6.155z" />
            </svg>
          </div>
        }
      </div>
    )
  }
}
