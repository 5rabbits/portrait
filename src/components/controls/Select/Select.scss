@import '~core/variables';
@import '~core/mixins';

.Select {
  position: relative;
  min-width: 240px;
  width: 100%;

  svg {
    fill: currentColor;
  }
}

.Select__control {
  @include reset-control;
  display: flex;
  align-items: center;
  height: 32px;
  padding: 0 16px;
  width: 100%;
  background-color: #fff;
  border: $default-input-border-color solid 1px;
  border-radius: $border-radius-medium-component;
}

.Select__placeholder,
.Select__value {
  @include text-ellipsis;
  margin-right: 16px;
  line-height: 1.25em;
}

.Select__placeholder {
  color: #B3BDC7;
}

.Select__arrow {
  margin-left: auto;
  display: flex;

  svg {
    height: 8px;
  }
}

.Select__dropdown {
  box-shadow: $interactive-shadow;
  position: absolute;
  right: 0px;
  left: 0;
  z-index: 9999999;
  background-clip: padding-box;
  background-color: #FFF;
  border: rgba(#000, 0.05) solid 1px;
  border-top-color: $default-input-border-color;
  border-bottom-left-radius: $border-radius-medium-component;
  border-bottom-right-radius: $border-radius-medium-component;
  overflow: hidden;
  margin-top: -1px;
}

.Select__emptyView {
  @include flex-center;
  padding: 16px;
  min-height: 80px;
  text-align: center;
}

.Select__menu {
  max-height: 160px;
  overflow: auto;
}

.Select__option {
  @include clickable;
  display: flex;
  align-items: center;
  height: 40px;
  padding: 0 16px;
}

.Select__option__text {
  display: -webkit-box;
  overflow: hidden;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  margin-right: auto;
  line-height: 1.25em;
  max-height: 2.5em;
  overflow: hidden;
}

.Select__option__icon {
  margin-left: 8px;
  display: flex;

  svg {
    height: 10px;
  }
}

.Select__input {
  display: flex;
  align-items: center;
  position: relative;
  border-bottom: #E7EAED solid 1px;

  input {
    height: 32px;
    border: none;
    width: 100%;
    padding: 0 40px 0 16px;
    outline: none;
  }
}

.Select__input__icon {
  @include reset-control;
  position: absolute;
  right: 16px;
  top: 50%;
  transform: translateY(-50%);
  color: $placeholder-color;
  pointer-events: none;
  display: flex;

  svg {
    height: 14px;
  }
}

.Select__input__icon--clickable {
  @include clickable;
  pointer-events: auto;
  color: $primary-color;
}

.Select__searchHighlight {
  font-weight: 700;
}

.Select__clearOption,
.Select__createOption {
  @include clickable;
  display: flex;
  align-items: center;
  padding: 0 16px;
  height: 48px;
  border-top: #E7EAED solid 1px;
}

.Select__clearOption {
  color: $placeholder-color;
}

.Select__createOption__icon {
  color: #BBB;
  margin-right: 8px;

  svg {
    height: 28px;
  }
}

.Select__createOption__textWrapper {
  flex: 1;
  min-width: 0;
}

.Select__createOption__label {
  font-size: 10px;
  font-weight: 300;
  line-height: 1.25em;
}

.Select__createOption__value {
  @include text-ellipsis;
  font-weight: 500;
  line-height: 1.25em;
}

.Select__clearOption--focused,
.Select__createOption--focused,
.Select__option--focused {
  background-color: #f7f7f7;
}

.Select__option--selected {
  background-color: #eee;
}

.Select--disabled {
  .Select__control {
    background-color: #f2f2f2;
    cursor: not-allowed;
  }

  .Select__arrow {
    opacity: 0.25;
  }
}

.Select--open {
  .Select__control {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
  }
}
