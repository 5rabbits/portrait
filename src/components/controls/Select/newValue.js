export default '$NEW_SELECT_VALUE$'
