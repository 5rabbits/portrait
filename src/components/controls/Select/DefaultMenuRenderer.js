/* eslint-disable */

import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import deburr from 'lodash/deburr'
import uniq from 'lodash/uniq'
import Option from './Option'

export default class DefaultMenuRenderer extends PureComponent {
  static propTypes = {
    color: PropTypes.string,
    focusedElement: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf(['bottom-control']),
    ]),
    menuRef: PropTypes.func.isRequired,
    onOptionClick: PropTypes.func.isRequired,
    onOptionMouseEnter: PropTypes.func.isRequired,
    optionRef: PropTypes.func.isRequired,
    optionRenderer: PropTypes.func.isRequired,
    optionSearchTerms: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })).isRequired,
    search: PropTypes.string.isRequired,
    value: PropTypes.any,
  }

  state = {
    fromIndex: null,
    toIndex: null,
  }

  componentDidMount() {
    this.updateViewport()
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.options !== nextProps.options) {
      this.setState({
        fromIndex: null,
        toIndex: null,
      })
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.options !== prevProps.options) {
      this.updateViewport()
    }
  }

  getOption = index => {
    const {
      color,
      focusedElement,
      onOptionClick,
      onOptionMouseEnter,
      optionRef,
      optionRenderer,
      optionSearchTerms,
      options,
      search,
      value,
    } = this.props
    const option = options[index]

    return (
      <Option
        color={color}
        focused={focusedElement === index}
        key={option.value}
        onClick={onOptionClick}
        onMouseEnter={onOptionMouseEnter}
        option={option}
        optionRef={optionRef}
        selected={value === option.value}
        style={{
          position: 'absolute',
          top: index * 40,
          left: 0,
          width: '100%',
        }}
        >
        {optionRenderer({
          cleanDiacritics: deburr,
          option,
          search,
          optionSearchTerms,
        })}
      </Option>
    )
  }

  handleScroll = () => {
    this.updateViewport()
  }

  updateViewport = () => {
    let fromIndex = Math.floor(this.menu.scrollTop / 40) - 1
    let toIndex = Math.ceil((this.menu.scrollTop + this.menu.offsetHeight) / 40) + 1

    fromIndex = Math.max(0, fromIndex)
    toIndex = Math.min(this.props.options.length, toIndex)

    this.setState({
      fromIndex,
      toIndex,
    })
  }

  menuRef = menu => {
    this.menu = menu
    this.props.menuRef(menu)
  }

  render() {
    const { options, value } = this.props
    const { fromIndex, toIndex } = this.state
    const indices = [0, options.length - 1]

    for (let i = fromIndex; i < toIndex; i += 1) {
      indices.push(i)
    }

    if (value) {
      const selectedIndex = options.findIndex(option => option.value === value)

      if (selectedIndex !== -1) {
        indices.push(selectedIndex)
      }
    }

    return (
      <div
        className="Select__menu"
        onScroll={this.handleScroll}
        ref={this.menuRef}
        >
        <div
          className="Select__innerMenu"
          style={{
            height: options.length * 40,
            position: 'relative',
          }}
          >
          {uniq(indices).map(this.getOption)}
        </div>
      </div>
    )
  }
}
