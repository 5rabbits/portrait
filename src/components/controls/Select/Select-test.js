import { React, toJSON, mount, shallow } from 'test/helper'
import deburr from 'lodash/deburr'
import Select from './Select'
import Option from './Option'
import defaultArrowRenderer from './defaults/arrowRenderer'
import defaultClearOptionRenderer from './defaults/clearOptionRenderer'
import defaultCreateOptionRenderer from './defaults/createOptionRenderer'
import defaultEmptyRenderer from './defaults/emptyRenderer'
import defaultInputPlaceholder from './defaults/inputPlaceholder'
import defaultInputRenderer from './defaults/inputRenderer'
import DefaultMenuRenderer from './defaults/menuRenderer'
import defaultNewOptionBuilder from './defaults/newOptionBuilder'
import defaultOptionRenderer from './defaults/optionRenderer'
import defaultOptionSearchTerms from './defaults/optionSearchTerms'
import defaultOptionsFilter from './defaults/optionsFilter'
import defaultPlaceholder from './defaults/placeholder'
import defaultPlaceholderRenderer from './defaults/placeholderRenderer'
import defaultSort from './defaults/sort'
import defaultValueRenderer from './defaults/valueRenderer'

describe('Select', () => {
  it('should render with minimum props', () => {
    const component = mount(<Select options={[]} />)

    expect(toJSON(component)).toMatchSnapshot()
  })

  describe('defaultArrowRenderer', () => {
    describe('if the select is open', () => {
      it('should return a caret up svg icon', () => {
        const arrow = shallow(defaultArrowRenderer({ isOpen: true }))

        expect(arrow.find('svg[data-icon-name="caret-up"]').exists()).toBe(true)
      })
    })

    describe('if the select is closed', () => {
      it('should return a caret down svg icon', () => {
        const arrow = shallow(defaultArrowRenderer({ isOpen: false }))

        expect(arrow.find('svg[data-icon-name="caret-down"]').exists()).toBe(true)
      })
    })
  })

  describe('defaultClearOptionRenderer', () => {
    describe('if focused', () => {
      it('should have class "Select__clearOption--focused"', () => {
        const clearOption = shallow(defaultClearOptionRenderer({
          focused: true,
        }))

        expect(clearOption.hasClass('Select__clearOption--focused')).toBe(true)
      })
    })

    describe('if not focused', () => {
      it('should not have class "Select__clearOption--focused"', () => {
        const clearOption = shallow(defaultClearOptionRenderer({
          focused: false,
        }))

        expect(clearOption.hasClass('Select__clearOption--focused')).toBe(false)
      })
    })

    it('should assign the specified onClick handler', () => {
      const onClick = jest.fn()
      const clearOption = shallow(defaultClearOptionRenderer({
        onClick,
      }))

      clearOption.simulate('click')

      expect(onClick).toHaveBeenCalled()
    })

    it('should assign the specified onMouseEnter handler', () => {
      const onMouseEnter = jest.fn()
      const clearOption = shallow(defaultClearOptionRenderer({
        onMouseEnter,
      }))

      clearOption.simulate('mouseenter')

      expect(onMouseEnter).toHaveBeenCalled()
    })
  })

  describe('defaultCreateOptionRenderer', () => {
    describe('if focused', () => {
      it('should have class "Select__createOption--focused"', () => {
        const clearOption = shallow(defaultCreateOptionRenderer({
          focused: true,
        }))

        expect(clearOption.hasClass('Select__createOption--focused')).toBe(true)
      })
    })

    describe('if not focused', () => {
      it('should not have class "Select__createOption--focused"', () => {
        const clearOption = shallow(defaultCreateOptionRenderer({
          focused: false,
        }))

        expect(clearOption.hasClass('Select__createOption--focused')).toBe(false)
      })
    })

    it('should assign the specified onClick handler', () => {
      const onClick = jest.fn()
      const clearOption = shallow(defaultCreateOptionRenderer({
        onClick,
      }))

      clearOption.simulate('click')

      expect(onClick).toHaveBeenCalled()
    })

    it('should assign the specified onMouseEnter handler', () => {
      const onMouseEnter = jest.fn()
      const clearOption = shallow(defaultCreateOptionRenderer({
        onMouseEnter,
      }))

      clearOption.simulate('mouseenter')

      expect(onMouseEnter).toHaveBeenCalled()
    })

    it('should render the specified label', () => {
      const clearOption = shallow(defaultCreateOptionRenderer({
        createOptionLabel: 'New test option',
      }))

      expect(clearOption.find('.Select__createOption__label').text()).toBe('New test option')
    })

    it('should render the specified value', () => {
      const clearOption = shallow(defaultCreateOptionRenderer({
        value: 'New option value',
      }))

      expect(clearOption.find('.Select__createOption__value').text()).toBe('New option value')
    })
  })

  describe('defaultEmptyRenderer', () => {
    describe('if there is a search', () => {
      it('should display "No results"', () => {
        const emptyView = shallow(defaultEmptyRenderer({
          search: 'Some search',
        }))

        expect(emptyView.text()).toBe('No results')
      })
    })

    describe('if there isn\'t a search', () => {
      it('should display "No options available"', () => {
        const emptyView = shallow(defaultEmptyRenderer({
          search: '',
        }))

        expect(emptyView.text()).toBe('No options available')
      })
    })
  })

  describe('defaultInputPlaceholder', () => {
    describe('if can create options', () => {
      it('should return "Write to search or create"', () => {
        const inputPlaceholder = defaultInputPlaceholder({
          canCreate: true,
        })

        expect(inputPlaceholder).toBe('Write to search or create')
      })
    })

    describe('if can\'t create options', () => {
      it('should return "Write to search"', () => {
        const inputPlaceholder = defaultInputPlaceholder({
          canCreate: false,
        })

        expect(inputPlaceholder).toBe('Write to search')
      })
    })
  })

  describe('defaultInputRenderer', () => {
    it('should bound an input to the specified onChange', () => {
      const onChange = jest.fn()
      const input = shallow(defaultInputRenderer({
        onChange,
      }))

      input.find('input[type="text"]').simulate('change')

      expect(onChange).toHaveBeenCalled()
    })

    it('should bound an input to the specified onKeyDown', () => {
      const onKeyDown = jest.fn()
      const input = shallow(defaultInputRenderer({
        onKeyDown,
      }))

      input.find('input[type="text"]').simulate('keydown')

      expect(onKeyDown).toHaveBeenCalled()
    })

    it('should set inputPlaceholder as the input placeholder', () => {
      const input = shallow(defaultInputRenderer({
        inputPlaceholder: 'Test placeholder',
      }))

      expect(input.find('input[type="text"]').prop('placeholder')).toBe('Test placeholder')
    })

    it('should set value as the input value', () => {
      const input = shallow(defaultInputRenderer({
        value: 'Some search',
      }))

      expect(input.find('input[type="text"]').prop('value')).toBe('Some search')
    })

    describe('it has a search with no results and can create options', () => {
      it('should display a button bound to onNewOptionClick', () => {
        const onNewOptionClick = jest.fn()
        const input = shallow(defaultInputRenderer({
          canCreate: true,
          hasResults: false,
          onNewOptionClick,
          value: 'Some search',
        }))
        const button = input.find({ onClick: onNewOptionClick })

        expect(button.exists()).toBe(true)
      })

      it('should display a plus svg icon', () => {
        const onNewOptionClick = jest.fn()
        const input = shallow(defaultInputRenderer({
          canCreate: true,
          hasResults: false,
          onNewOptionClick,
          value: 'Some search',
        }))

        const button = input.find({ onClick: onNewOptionClick })

        expect(button.find('svg[data-icon-name="plus"]').exists()).toBe(true)
      })
    })

    describe('it has no search', () => {
      it('should display a search svg icon', () => {
        const input = shallow(defaultInputRenderer({
          value: '',
        }))

        expect(input.find('svg[data-icon-name="search"]').exists()).toBe(true)
      })
    })

    describe('it has a search with results', () => {
      it('should display a search svg icon', () => {
        const input = shallow(defaultInputRenderer({
          hasResults: true,
          value: 'Some search',
        }))

        expect(input.find('svg[data-icon-name="search"]').exists()).toBe(true)
      })
    })

    describe('it has a search with no results but it can\'t create options', () => {
      it('should display a search svg icon', () => {
        const input = shallow(defaultInputRenderer({
          hasResults: false,
          value: 'Some search',
          canCreate: false,
        }))

        expect(input.find('svg[data-icon-name="search"]').exists()).toBe(true)
      })
    })
  })

  describe('defaultMenuRenderer', () => {
    const renderMenu = (props = {}) =>
      mount(
        <DefaultMenuRenderer
          focusedElement={null}
          menuRef={jest.fn()}
          onOptionClick={jest.fn()}
          onOptionMouseEnter={jest.fn()}
          optionRef={jest.fn()}
          optionRenderer={defaultOptionRenderer}
          optionSearchTerms={defaultOptionSearchTerms}
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 2', value: 2 },
            { label: 'Option 3', value: 3 },
            { label: 'Option 4', value: 4 },
            { label: 'Option 5', value: 5 },
            { label: 'Option 6', value: 6 },
            { label: 'Option 7', value: 7 },
            { label: 'Option 8', value: 8 },
            { label: 'Option 9', value: 9 },
          ]}
          search=""
          value={null}
          {...props}
        />,
      )

    it('should render()', () => {
      const menu = renderMenu()

      expect(toJSON(menu)).toMatchSnapshot()
    })

    describe('when it receives new options', () => {
      it('should render the new options', () => {
        const menu = renderMenu()
        const options = [
          { label: 'Option 1', value: 1 },
          { label: 'Option 2', value: 2 },
        ]

        const snapshotBefore = toJSON(menu)

        menu.setProps({ options })

        const snapshotAfter = toJSON(menu)

        expect(snapshotBefore).not.toEqual(snapshotAfter)
      })
    })

    describe('when it receives the same options', () => {
      it('should render the same', () => {
        const menu = renderMenu()
        const snapshotBefore = toJSON(menu)

        menu.setProps({ options: menu.prop('options') })

        const snapshotAfter = toJSON(menu)

        expect(snapshotBefore).toEqual(snapshotAfter)
      })
    })

    it('should update the viewport when scrolling on the menu', () => {
      const menu = renderMenu()
      const instance = menu.instance()

      instance.updateViewport = jest.fn()

      menu.simulate('scroll')

      expect(instance.updateViewport).toHaveBeenCalled()
    })

    describe('if it has a value', () => {
      it('should always render the selected option', () => {
        const menu = renderMenu({
          value: 2,
        })
        const option = menu.findWhere(child =>
          child.is(Option) && child.key() === '2',
        )

        expect(option.exists()).toBe(true)
      })

      describe('if the selected option is not defined', () => {
        it('should not throw', () => {
          expect(() => {
            renderMenu({ value: 123123 })
          }).not.toThrow()
        })
      })
    })
  })

  describe('defaultNewOptionBuilder', () => {
    it('should return a valid option with the specified label and value', () => {
      const newOption = defaultNewOptionBuilder({
        label: 'Some text',
        value: 'Some value',
      })

      expect(newOption).toEqual({
        label: 'Some text',
        value: 'Some value',
      })
    })
  })

  describe('defaultOptionRenderer', () => {
    describe('if it has a search', () => {
      it('should build a highlighted label', () => {
        const renderedOption = shallow(defaultOptionRenderer({
          cleanDiacritics: deburr,
          option: { label: 'Test label' },
          optionSearchTerms: defaultOptionSearchTerms,
          search: 'te abe',
        }))

        expect(toJSON(renderedOption)).toMatchSnapshot()
      })
    })

    describe('if it hasn\'t a search', () => {
      it('should return the option label', () => {
        const renderedOption = defaultOptionRenderer({
          option: { label: 'Test label' },
          search: '',
        })

        expect(renderedOption).toBe('Test label')
      })
    })
  })

  describe('defaultOptionSearchTerms', () => {
    it('should return matching info for the given option', () => {
      const terms = defaultOptionSearchTerms({
        cleanDiacritics: deburr,
        option: { label: 'Test label' },
        search: 'te abe',
      })

      expect(terms).toEqual([
        {
          text: 'Te',
          matches: true,
          fromIndex: 0,
          toIndex: 2,
        },
        {
          text: 'st l',
          matches: false,
          fromIndex: 2,
          toIndex: 6,
        },
        {
          text: 'abe',
          matches: true,
          fromIndex: 6,
          toIndex: 9,
        },
        {
          text: 'l',
          matches: false,
          fromIndex: 9,
          toIndex: 10,
        },
      ])
    })
  })

  describe('defaultOptionsFilter', () => {
    describe('if it has a search', () => {
      it('should return the options that match the search without casing and diacritics', () => {
        const filtered = defaultOptionsFilter({
          cleanDiacritics: deburr,
          search: 'ó i 1',
          options: [
            { label: 'Optïòn 11' },
            { label: 'óption 27' },
            { label: 'Ôption 10' },
            { label: 'Option 23' },
          ],
        })

        expect(filtered).toEqual([
          { label: 'Optïòn 11' },
          { label: 'Ôption 10' },
        ])
      })
    })

    describe('if it hasn\'t a search', () => {
      it('should return the options without processing them', () => {
        const options = []
        const filtered = defaultOptionsFilter({
          search: '',
          options,
        })

        expect(filtered).toBe(options)
      })
    })
  })

  describe('defaultPlaceholder', () => {
    describe('if can create options', () => {
      it('should return "Select or create an option"', () => {
        const inputPlaceholder = defaultPlaceholder({
          canCreate: true,
        })

        expect(inputPlaceholder).toBe('Select or create an option')
      })
    })

    describe('if can\'t create options', () => {
      it('should return "Select an option"', () => {
        const inputPlaceholder = defaultPlaceholder({
          canCreate: false,
        })

        expect(inputPlaceholder).toBe('Select an option')
      })
    })
  })

  describe('defaultPlaceholderRenderer', () => {
    it('should render the specified placeholder', () => {
      const renderedPlaceholder = shallow(defaultPlaceholderRenderer({
        placeholder: 'Some placeholder',
      }))

      expect(renderedPlaceholder.text()).toBe('Some placeholder')
    })
  })

  describe('defaultSort', () => {
    it('should sort all options without casing and diacritics', () => {
      const sorted = defaultSort({
        cleanDiacritics: deburr,
        options: [
          { label: 'a3' },
          { label: 'b2' },
          { label: 'Á1' },
        ],
      })

      expect(sorted).toEqual([
        { label: 'Á1' },
        { label: 'a3' },
        { label: 'b2' },
      ])
    })
  })

  describe('defaultValueRenderer', () => {
    it('should render the option label', () => {
      const renderedValue = shallow(defaultValueRenderer({
        option: { label: 'Test option' },
      }))

      expect(renderedValue.text()).toBe('Test option')
    })
  })
})
