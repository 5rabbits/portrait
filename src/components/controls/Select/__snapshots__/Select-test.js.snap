// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select defaultMenuRenderer should render() 1`] = `
<DefaultMenuRenderer
  focusedElement={null}
  menuRef={[Function]}
  onOptionClick={[Function]}
  onOptionMouseEnter={[Function]}
  optionRef={[Function]}
  optionRenderer={[Function]}
  optionSearchTerms={[Function]}
  options={
    Array [
      Object {
        "label": "Option 1",
        "value": 1,
      },
      Object {
        "label": "Option 2",
        "value": 2,
      },
      Object {
        "label": "Option 3",
        "value": 3,
      },
      Object {
        "label": "Option 4",
        "value": 4,
      },
      Object {
        "label": "Option 5",
        "value": 5,
      },
      Object {
        "label": "Option 6",
        "value": 6,
      },
      Object {
        "label": "Option 7",
        "value": 7,
      },
      Object {
        "label": "Option 8",
        "value": 8,
      },
      Object {
        "label": "Option 9",
        "value": 9,
      },
    ]
  }
  search=""
  value={null}
>
  <div
    className="Select__menu"
    onScroll={[Function]}
  >
    <div
      className="Select__innerMenu"
      style={
        Object {
          "height": 360,
          "position": "relative",
        }
      }
    >
      <Option
        focused={false}
        onClick={[Function]}
        onMouseEnter={[Function]}
        option={
          Object {
            "label": "Option 1",
            "value": 1,
          }
        }
        optionRef={[Function]}
        selected={false}
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
      >
        <div
          className="Select__option"
          onClick={[Function]}
          onMouseEnter={[Function]}
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": "100%",
            }
          }
        >
          <div
            className="Select__option__text"
          >
            Option 1
          </div>
        </div>
      </Option>
      <Option
        focused={false}
        onClick={[Function]}
        onMouseEnter={[Function]}
        option={
          Object {
            "label": "Option 9",
            "value": 9,
          }
        }
        optionRef={[Function]}
        selected={false}
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": 320,
            "width": "100%",
          }
        }
      >
        <div
          className="Select__option"
          onClick={[Function]}
          onMouseEnter={[Function]}
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 320,
              "width": "100%",
            }
          }
        >
          <div
            className="Select__option__text"
          >
            Option 9
          </div>
        </div>
      </Option>
    </div>
  </div>
</DefaultMenuRenderer>
`;

exports[`Select defaultOptionRenderer if it has a search should build a highlighted label 1`] = `
<div>
  <span
    className="Select__searchHighlight"
  >
    Te
  </span>
  st l
  <span
    className="Select__searchHighlight"
  >
    abe
  </span>
  l
</div>
`;

exports[`Select should render with minimum props 1`] = `
<Select
  arrowRenderer={[Function]}
  autoFocus={false}
  canCreate={false}
  clearOptionRenderer={[Function]}
  clearText="Leave without value"
  clearValue={null}
  clearable={false}
  createOptionLabel="Create new option"
  createOptionRenderer={[Function]}
  defaultNewOptionLabel=""
  defaultValue={null}
  disabled={false}
  emptyRenderer={[Function]}
  inputPlaceholder={[Function]}
  inputRenderer={[Function]}
  menuRenderer={[Function]}
  newOptionBuilder={[Function]}
  onChange={[Function]}
  optionRenderer={[Function]}
  optionSearchTerms={[Function]}
  options={Array []}
  optionsFilter={[Function]}
  placeholder={[Function]}
  placeholderRenderer={[Function]}
  searchable={true}
  sort={[Function]}
  valueRenderer={[Function]}
>
  <div
    className="Select"
  >
    <button
      autoFocus={false}
      className="Select__control"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="button"
    >
      <div
        className="Select__placeholder"
      >
        Select an option
      </div>
      <div
        className="Select__arrow"
      >
        <svg
          data-icon-name="caret-down"
          viewBox="0 0 400 400"
        >
          <path
            d="M3.625 110.875l2.25 2.625 170 195.625c5.75 6.625 14.375 10.75 24 10.75s18.25-4.25 24-10.75l169.875-195.25 2.875-3.25C398.75 107.5 400 103.75 400 99.75 400 88.875 390.75 80 379.25 80H20.75C9.25 80 0 88.875 0 99.75c0 4.125 1.375 8 3.625 11.125z"
          />
        </svg>
      </div>
    </button>
  </div>
</Select>
`;
